x/interchainstaking/keeper/callbacks_test.go

```diff
@@ -1607,5 +1647,163 @@ func (suite *KeeperTestSuite) TestDelegationAccountBalanceCallback() {
+
+			// initialize the callback
+			err := keeper.DepositTxCallback(qckApp.InterchainstakingKeeper, ctx, tc.txWithProofbz, icqtypes.Query{ChainId: tc.chainID, Request: resDataBz})
+			if tc.expectErr {
+				suite.Error(err)
+			} else {
+				suite.NoError(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestCheckTMHeaderForZone() {
+	testCases := []struct {
+		name        string
+		expectedErr bool
+		changeCtx   func(qck *app.Quicksilver, ctx *sdk.Context, txRes *icqtypes.GetTxWithProofResponse, zoneB *icstypes.Zone)
+	}{
+		{
+			name:        "Check TM Header for Zone successful",
+			expectedErr: false,
+			changeCtx:   nil,
+		},
+		{
+			name:        "Check TM Header for Zone failed: unable to fetch client state",
+			expectedErr: true,
+			changeCtx: func(qck *app.Quicksilver, ctx *sdk.Context, txRes *icqtypes.GetTxWithProofResponse, zone *icstypes.Zone) {
+				zone.ConnectionId = "connection-1"
+			},
+		},
+		{
+			name:        "Check TM Header for Zone failed: unable to fetch consensus state on trusted height",
+			expectedErr: true,
+			changeCtx: func(qck *app.Quicksilver, ctx *sdk.Context, txRes *icqtypes.GetTxWithProofResponse, zone *icstypes.Zone) {
+				txRes.Header.TrustedHeight = ibctypes.Height{RevisionNumber: 1, RevisionHeight: 24072000}
+			},
+		},
+		{
+			name:        "Check TM Header for Zone failed: unable to verify TM Header",
+			expectedErr: true,
+			changeCtx: func(qck *app.Quicksilver, ctx *sdk.Context, txRes *icqtypes.GetTxWithProofResponse, zone *icstypes.Zone) {
+				*ctx = ctx.WithBlockTime(ctx.BlockHeader().Time.Add(time.Hour * 24 * 8))
+			},
+		},
+		{
+			name:        "Check TM Header for Zone failed: unable to marshal proof",
+			expectedErr: true,
+			changeCtx: func(qck *app.Quicksilver, ctx *sdk.Context, txRes *icqtypes.GetTxWithProofResponse, zone *icstypes.Zone) {
+				txRes.Proof = &types.TxProof{}
+			},
+		},
+		{
+			name:        "Check TM Header for Zone failed: unable to verify TM Header from wrong data hash",
+			expectedErr: true,
+			changeCtx: func(qck *app.Quicksilver, ctx *sdk.Context, txRes *icqtypes.GetTxWithProofResponse, zone *icstypes.Zone) {
+				txRes.Header.Header.DataHash = []byte("wrong data hash")
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.name), func() {
+			// setup app and counter chain ctx
+			qckApp, ctx := suite.setupIbc()
+
+			// get chain B
+			zone, _ := qckApp.InterchainstakingKeeper.GetZone(ctx, suite.chainB.ChainID)
+
+			// get the tx from fixture
+			txWithProofBz := decodeBase64NoErr(localDepositTxFixture)
+			txRes := icqtypes.GetTxWithProofResponse{}
+			_ = qckApp.InterchainQueryKeeper.IBCKeeper.Codec().Unmarshal(txWithProofBz, &txRes)
+
+			// setup ctx for testcase
+			if tc.expectedErr {
+				tc.changeCtx(qckApp, &ctx, &txRes, &zone)
+			}
+
+			err := qckApp.InterchainstakingKeeper.CheckTMHeaderForZone(ctx, &zone, txRes)
+			if tc.expectedErr {
+				fmt.Println(err)
```

Comment:
This used to debug right @ hoangdv2429? Overall everything is okay, need to fix the conflict then we can merge